### ISSUE TITLE
Bad Gateway and Config Linting

### ISSUE THREAD

**Issue by aranw** [facebookarchive/heroku-buildpack-hhvm#15]
When trying to use this buildpack with Dokku I came across two problems.

The first when using default setup I got a 502 Bad Gateway error.

I then tried to add a custom `config.hdf` to change the settings to see if that would help but what ever I put I got a `Braces don't match. Found ${braces_count} more than expected`

**Comment by ClaudiuC** [comment]
For the 502 Bad Gateway error, can you please add more details? How did you use it with Dokku? (I honestly haven't tried Dokku before, so repro steps would help).

Also, it looks like the .hdf file you're using broke the linter, mind sharing it so I can test and fix please? We have tests for the linter at: https://github.com/hhvm/heroku-buildpack-hhvm/tree/master/bin/linter/tests. It's weird that the error message actually says `${braces_count}`... What OS are you running this on?

**Comment by aranw** [comment]
Sorry for confusion it didn't have the `${braces_count}` in it, I just copied that from the lint tester for reference. Also was trying to see what could cause the bug. the number was 1.

As for setup in regards to Dokku, I just point it to the Buildpack. Similar to how Heroku works. 

The config file I was using just set the SourceRoot and Document

```
Server {
    SourceRoot = public
    DefaultDocument = index.php
}
```

I copied the code from other config files, I even copied the one out of the buildpack and tried that. Still no luck though.

As for environment don't really know exactly what is used as it boots into a Docker container, and installs whatever the Buildpack tells it to install.

**Comment by ClaudiuC** [comment]
Thanks for the info! I doubt Dokku will work unfortunately, at least not with the current buildpack version. From the Dokku readme:

```
Assumes Ubuntu 13 or 12.04 x64 right now. Ideally have a domain ready to point to your host.   It's designed for and is probably best to use a fresh VM. The bootstrapper will install everything it   needs.
```

Heroku dynos actually run Ubuntu 10.04 and we download an HHVM binary built for Ubuntu 10.04. 

That being said, I'll take a look and see:

1) Why the .hdf fails - there's nothing obvious that I can see
2) Think about supporting Dokku, we have prebuilt binaries for 12.04 and 13.04 afaik so it could be done. 

Thanks for the report!

**Comment by aranw** [comment]
Ah ok, well just let me know if you need it testing and I'll give it a try for you.

**Comment by henrikbjorn** [comment]
Is it possible to get the host os type when compiling? and use the appropiate binary? This would also fix issues if Heroku ever decides to upgrade their platform.

**Comment by aranw** [comment]
Was just seeing how CHH handles it in his PHP Buildpack but I can't seem to see anything that is detecting the OS and downloading specific binaries for the OS type. https://github.com/CHH/heroku-buildpack-php/blob/master/bin/compile

Heroku Buildpacks should work out the of box on Dokku, which surprises me that OS versions haven't been considered cause that's obviously a big issue for support.

**Comment by ClaudiuC** [comment]
Yup, will have to look into it. Theoretically we could not make assumptions about the environment and download the correct binary, just need to tinker with it. In fairness, right now, the buildpack is pretty "tightly coupled" with what Heroku is currently running on their dynos, we should be able to do better. 

@henrikbjorn To your point, we should be able to use `uname`:

```
vagrant@vagrant-ubuntu-raring-64:~/hacktest$ uname -a
Linux vagrant-ubuntu-raring-64 3.11.0-18-generic #32-Ubuntu SMP Tue Feb 18 21:11:14 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux
```

And then something platform specific I guess:

```
vagrant@vagrant-ubuntu-raring-64:~/hacktest$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 13.10
Release:    13.10
Codename:   saucy
```

Edit: Here's a complete script that does something similar: https://github.com/coto/server-easy-install/blob/master/lib/core.sh

**Comment by hipsterjazzbo** [comment]
Wasn't sure whether to start a new issue, but my `config.hdf` is also failing linting, complaining about mismatched braces. 

Here's the error:

```
-----> Sourcing config.hdf
-----> Found custom config.hdf, linting
-----> Braces don't match. Found 1 more than expected.
-----> Aborting - can't push bad config

 !     Push rejected, failed to compile PHP (HHVM) app
```

and here the `config.hdf`

```
Server {
    Port = 80
    SourceRoot = /app/public/
}

VirtualHost {
    * {
        Pattern = .*
        RewriteRules {
            * {
                pattern = .?

                # app bootstrap
                to = index.php

                # append the original query string
                qsa = true
            }
        }
    }
}

StaticFile {
    Extensions {
        css = text/css
        gif = image/gif
        html = text/html
        jpe = image/jpeg
        jpeg = image/jpeg
        jpg = image/jpeg
        png = image/png
        tif = image/tiff
        tiff = image/tiff
        txt = text/plain
    }
}
```
